### DISCUSSION TITLE
일급 컬렉션이 무엇이고 왜 사용하는지?

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#27] [Java]
.

**Comment by gyungchan-jo** [comment]
Collection을 Wrapping하면서, 그 외 다른 멤버 변수가 없는 상태를 일급 컬렉션이라 한다.

일급 컬렉션을 사용하는 이유:
1. 비지니스에 종속적인 자료구조를 직접 만들 수 있어 이후 발생할 문제를 최소화시킬 수 있다.
2. Collection의 불변성을 보장할 수 있다.
3. 상태와 행위를 한 곳에서 관리할 수 있다.
4. 컬렉션에 이름을 붙일 수 있어 관리하기 용이하다.

**Comment by le2sky** [comment]
콜렉션을 감싸며, 이외에 다른 필드를 가지고 있지 않는 클래스를 일급 컬렉션이라고 합니다.
사용하는 이유는 다음과 같습니다.

컬렉션을 감싼 구조에 도메인 로직을 포함할 수 있습니다.
불필요한, 컬렉션의 API를 외부에 노출하지 않도록 할 수 있습니다.
일급 컬렉션의 역할에 맞는 명확한 이름을 지어줄 수 있습니다.

**Comment by kunsanglee** [comment]
일급 컬렉션(First-Class Collection)이란, 내부에 단 하나의 컬렉션을 감싸는 클래스로, 이 클래스에서 컬렉션과 관련된 비즈니스 로직을 관리하는 패턴을 말합니다. 이 개념은 객체 지향 프로그래밍에서 값과 로직을 함께 캡슐화하여, 코드의 유지보수성과 가독성을 높이는 데 도움을 줍니다.

일급 컬렉션은 컬렉션에 관련된 책임만 가지고 있기 때문에 단일 책임 원칙을 지킬 수 있습니다. 
또한 일급 컬렉션을 사용하는 클래스가 내부에 컬렉션의 검증, 비즈니스 로직에 대한 책임을 가지지 않습니다. 
즉, 컬렉션과 관련된 로직이 캡슐화되어 외부의 불필요한 접근을 제한하여 디미터의 법칙을 지킬 수 있습니다.

**Comment by GIVEN53** [comment]
컬렉션 하나만 멤버 변수
불변성, 도메인 로직, 이름 부여, 캡슐화
